net: added CoAP RD client for simple node registration

haukepetersen opened the pull request:
~~based on #7402~~
also needs
- ~~https://github.com/kaspar030/sock/pull/16~~
- ~~https://github.com/kaspar030/sock/pull/17~~

This PR provides a CoAP resource directory (RD) client implementation, that uses the simplified client registration as described in section 5.3.2 in [draft-ietf-core-resource-directory-11](https://tools.ietf.org/pdf/draft-ietf-core-resource-directory-11.pdf). The client implementation can be used in two different ways:

1. standalone (`USEMODULE += rdcli_simple_standalone`): in this mode the module is running in its own thread, which is started automatically on system startup via `auto-init`. In this mode, nothing has to be done by the user.
2. 'manual' (`USEMODULE += rdcli_simple`): in this mode, the user has to take care of calling `rdcli_simple_register()` for initial registration and then periodically or on certain events to refresh the registration and to reset the lifetime counter in the resource directory.

The `rdcli_simple` client is statically configured through the `sys/net/include/rdcli_config.h` header.

smlng commented:
just a quick one on first sight: I find the naming (`rdcli`) a bit vague, i.e. its the _CoAP Resource Directory_, hence I would expect `coap` to be part of the name/prefix?

haukepetersen commented:
but then the name gets too long... Also, I just saw, the official name is CoRE Resource Directory, so the natural name would be `corerdcli`?! But a quick google search for `resource directory` points very reliable to the `CoRE RD`, so IMHO it is fine to omit the `CoRE` from the name of the module and stick to `rdcli`. I could however change the name of the example project to something more meaningful...

smlng commented:
mhm, I thought RD is intrinsically connected to/with CoAP - but as I just read, its not CoAP is just one way to access resources. So never mind, leave naming as is.

haukepetersen commented:
awesome. Would you be inclined to maybe have a look at the underlying PR by any chance?!

bergzand commented on sys/net/application_layer/rdcli_simple/rdcli_simple_standalone.c:
```diff
@@ -0,0 +1,54 @@
+    (void)arg;
+
+    /* wait some seconds to give the address configuration some time to settle */
+    xtimer_sleep(3);
```
Isn't more useful to have a define for this delay? Are there situations where 3 seconds is insufficient?

haukepetersen replied:
Probably yes. Will add this.

haukepetersen commented:
startup delay is now configurable. Still waiting for kaspar030/sock#17, though.

haukepetersen commented:
kaspar030/sock#17 was merged, so rebased and updated the nanocoap pkg

haukepetersen commented:
Did some additional enhancements and fixes:
- removed all dependencies for hwrng from example application
- improved resources in example
- fixed doxygen in many places
- fixed gcoap to accept NULL as response handler
- renamed example to `rdcli_simple`

bergzand commented:
Is there any advantage to having the separate gcoap RD examples instead of integrating them in the gcoap example? As a test, I think it is useful to have a minimal CoAP RD test, but I think it would be better to have the example integrated in the gcoap example.

haukepetersen commented:
Yes, so far it's separation of concerns... The example of introduced by this PR is to demonstrate the simplified registration as most likely used by very minimal setups, and this is IMHO best demonstrated by a minimal demo application. I would then rather integrate the full RD client as part of the gcoap example...

bergzand commented on sys/net/application_layer/rdcli_simple/rdcli_simple.c:
```diff
@@ -0,0 +1,60 @@
+        return res;
+    }
+    /* finish, we don't have any payload */
+    ssize_t len = gcoap_finish(&pkt, 0, COAP_FORMAT_TEXT);
```
Looking at the [rfc draft section 5.3.2](https://tools.ietf.org/html/draft-ietf-core-resource-directory-11#section-5.3.2), this should be `COAP_FORMAT_LINK`.

haukepetersen commented on examples/rdcli_simple/Makefile:
```diff
@@ -0,0 +1,34 @@
+USEMODULE += shell_commands
+USEMODULE += ps
+
+FEATURES_REQUIRED += periph_hwrng
```
ups, leftover from a previous example, will drop

haukepetersen commented on Makefile.dep:
```diff
@@ -652,6 +652,22 @@ USEPKG += nanocoap
+
+ifneq (,$(filter rdcli_common,$(USEMODULE)))
+  USEMODULE += luid
+  USEMODULE += fmt
```
yes

haukepetersen commented:
@bergzand addressed your comments, thanks for the review!

smlng commented:
please rebase, also `nanocoap` is now part of RIOT, i.e. no external package anymore.

haukepetersen commented on examples/rdcli_simple/main.c:
```diff
@@ -0,0 +1,89 @@
+ * @{
+ *
+ * @file
+ * @brief       CoRE RD simple registration test application
```
done

haukepetersen commented:
rebased and fixed and adapted to latest nanocoap/gcoap changes.

haukepetersen commented:
I just cut out the `gcaop` changes and put them into a separate PR: #8800

So this PR is now based on #8800

bergzand commented on examples/rdcli_simple/Makefile:
```diff
@@ -0,0 +1,32 @@
+# For debugging and demonstration purposes, we limit the lifetime to the minimal
+# allowed value of 60s (see draft-ietf-core-resource-directory-11, Table 2)
+CFLAGS += -DRDCLI_LT=60
+
```
Although the README already states how to override the default RD server, I think it would be convenient to have them configurable here, if only as an example on how to override them and to remind people to configure to what suits them.

Maybe something like:
```
# Uncomment the following lines to override the default CoRE RD:
#RDCLI_SERVER="IPV6_ADDR_ALL_ROUTERS_LINK_LOCAL"
#CFLAGS=-DRDCLI_SERVER_ADDR=$(RDCLI_SERVER)
```

bergzand commented on sys/net/application_layer/rdcli_simple/rdcli_simple_standalone.c:
```diff
@@ -0,0 +1,54 @@
+        if (res < 0) {
+            LOG_ERROR("[rdcli_simple] error: unable to trigger registration\n");
+        }
+        xtimer_sleep(RDCLI_UPDATE_INTERVAL);
```
I think it would be good to have a distinction here between a sleep interval when successfully registered and when triggering the registration failed. With this, if a registration fails, the next registration occurs at the exact moment the registration expires. With the default settings, if the node doesn't have connectivity within 3 seconds, it takes half a day before it registers.

I don't consider this blocking for now, it could always be done in a follow up PR.

bergzand replied:
I missed that it sends confirmable messages, so this is already handled by the gcoap code :)

haukepetersen replied:
:-)

haukepetersen commented:
reabsed and added the RD server address configuration to the example's Makefile.

@bergzand can I squash?

bergzand commented:
ACK

> @bergzand can I squash?

Yes, I'd say this is ready to get merged unless somebody else wants to start nitpicking over it at this point :)

kaspar030 commented on sys/net/application_layer/rdcli_simple/rdcli_simple_standalone.c:
```diff
@@ -0,0 +1,54 @@
+#define PRIO                (THREAD_PRIORITY_MAIN - 1)
+#define TNAME               "rdcli_simple"
+
+char stack[STACKSIZE];
```
static

kaspar030 commented on sys/net/application_layer/rdcli_common/rdcli_common.c:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    luid_get(luid, sizeof(luid));
+    for (unsigned i = 0; i < sizeof(luid); i++) {
```
there's a fmt_bytes_hex now

haukepetersen replied:
fixed.

kb2ma commented:
@haukepetersen, @bergzand: What did you use as an RD server for development/testing?

kb2ma commented:
> @haukepetersen, @bergzand: What did you use as an RD server for development/testing?

Never mind. I see Hauke's [comment](https://github.com/RIOT-OS/RIOT/pull/7428#issuecomment-320188393) in #7428 on @pekkanikander's JavaScript-based server.

kb2ma commented:
I haven't actually tested the implementation but have a couple of comments. Overall, I think this work is a nice addition. A resource directory seems like a worthwhile service to reduce the burden on a constrained network.

I understand the need for flexibility at this point in the implementation. Let the user decide whether to run the registration themselves or in an external thread. However, registration management seems like an infrastructure task rather than something an application developer should care about in detail. gcoap itself seems well suited to this work. It runs in its own thread and is designed as a kind of messaging hub. In the future, it can provide resource registration as an optional service.

gcoap already has the gcoap_register_listener() function to trigger registration. Then it can use its thread message handler loop to react to a timer to re-register based on lifetime. Maybe this work could be the beginning of a plug-in mechanism for optional CoAP features.

Just to be clear, I'm envisioning gcoap-based registration as future work, and not blocking this PR. :-)

haukepetersen commented:
rebased and squashed.

@kaspar030 fixed your remarks and merged them right in

@kb2ma Thanks for your thoughts! I think what you sketch makes a lot of sense, and reusing the gcoap's thread for this kind of tasks seems like a very valid approach. Being able to integrate this RD implementation into a 'larger' scope is exactly the reason, why I factored out the thread-stuff into an optional sub-module. Once this PR is merged, I will bring the 'full' RD client PR (#7428) back in a good shape. I would think that that PR is suited well to filter out the exact requirements on gcoap for running additionally 'CoAP based services'. So lets see were we can take that.

haukepetersen commented:
fixed header guards and sqqashed.

bergzand commented:
@haukepetersen Could you please add the failed boards from Murdock to the examples `BOARD_INSUFFICIENT_MEMORY` list? Feel free to keep it squashed :)

haukepetersen commented:
done

haukepetersen commented:
Travis seems to have a point here, do not merge yet!

haukepetersen commented:
nope, false alarm, defining a custom value for `RDCLI_EP` does work as expected, I merely had some trouble escaping the `"` chars locally.

So until somebody can explain why `Travis` complains about `RDCLI_EP` being `void`, I'd say its save to merge this PR...

haukepetersen commented:
@bergzand what do you say, should we merge?

bergzand commented:
Murdock is not complaining about it, so I blame the older cppcheck from travis.
